cvat-sdk/cvat_sdk/core/uploading.py

```diff
@@ -180,13 +180,25 @@ def upload_file(
         # query params are used only in the extra messages
         assert meta["filename"]
 
+        if pbar is None:
+            pbar = NullProgressReporter()
+
+        file_size = filename.stat().st_size
+
         self._tus_start_upload(url, query_params=query_params)
-        real_filename = self._upload_file_data_with_tus(
-            url=url, filename=filename, meta=meta, pbar=pbar, logger=logger
-        )
+        with self._uploading_task(pbar, file_size):
+            real_filename = self._upload_file_data_with_tus(
+                url=url, filename=filename, meta=meta, pbar=pbar, logger=logger
+            )
         query_params["filename"] = real_filename
         return self._tus_finish_upload(url, query_params=query_params, fields=fields)
 
+    @staticmethod
+    def _uploading_task(pbar: ProgressReporter, total_size: int) -> ContextManager[None]:
+        return pbar.task(
+            total=total_size, desc="Uploading data", unit_scale=True, unit="B", unit_divisor=1024
```

Comment:
I see several problems with this change:

- It supposes that the progress reporter is tqdm
- There is no way for the user to implement their reporter correctly, because the parameters used in SDK are not specified anywhere, they are just passed, and they are similar everywhere, hopefully
- The suffix `2` is quite cryptic and tells nothing about the differences, probably it can be replaced with `Lazy` or `Deferred`

Reply:
> * The implementation supposes that the progress reporter is tqdm
> * The tqdm has quite a broad list of parameters, it will be hard to implement a custom reporter correctly

That's not quite true, because a reporter doesn't have to support every parameter that tqdm supports. It can simply use whatever parameters it recognizes, and ignore the rest. A barebones reporter doesn't have to use _any_ information other than `total` (maybe not even `total`...).

That said, I think I can improve this by explicitly listing the parameters that we actually use right now, and adding a note saying that any kwargs should be ignored. That way, users have more guidance on which parameters they should pay attention to, and there's a way for us to introduce non-tqdm parameters in the future if we need them.

> The suffix `2` is quite cryptic and tells nothing about the differences, probably it can be replaced with Lazy or Deferred

The only thing that the `2` is supposed to communicate is that it's better than 1. Which, to me, seems like the main thing that users need to know in this case. If you really want to, I can use `Deferred` instead.

Reply:
> Sounds good. I think it's the right direction.

Okay, I did it.

> I won't claim that deferred is the best variant, but 2 feels like "TODO" for me - like if we couldn't find a good name and just left it for future.

Mind you, the good name is `TqdmProgressReporter` - I just couldn't use it because it was already taken. 🙂 

Well, I renamed it to `DeferredTqdmProgressReporter` anyway.